### ISSUE TITLE
[BUG] Z-Axis movement doesn't stop immediately when Z-probe triggers

### ISSUE THREAD

**Issue by bbrandst** [MarlinFirmware/Marlin#22988]
### Did you test the latest `bugfix-2.0.x` code?

Yes, and the problem still exists.

### Bug Description

I have converted my Anycubic i3 Mega S to Z-axis probing directly via the nozzle (using a strain gauge).
Everything works perfectly fine except that the Z movement doesn't stop immediately once the probe triggers. Instead, the nozzle is driving further down into the print bed for about 0.4 mm; I can clearly see that the spring-loaded bed gives way. This also results in wrong readings during bed leveling.
For instance, I verified that at coordinate point X10/Y10 my probe triggers at Z=-0.04mm when moving down the print head manually via G1. But when running G29, the Z level for this point is reported to be -0.48.
So far I was not able to find any configuration option affecting this behavior, nor could I find a potential cause by inspecting the code.

### Bug Timeline

I also tried with Marlin 1.1.9 and the behavior was the same.

### Expected behavior

I expect the Z down-movement to stop immediately once the probe triggers.

### Actual behavior

The nozzle continues moving down for an additional ~0.4 mm (or ~0.4 s) after the probe has triggered.

### Steps to Reproduce

Run G29 on a printer using NOZZLE_AS_PROBE

### Version of Marlin Firmware

bugfix-2.0.x

### Printer model

Anycubic i3 Mega S

### Electronics

Stock electronics

### Add-ons

Added z-probe

### Bed Leveling

ABL Bilinear mesh

### Your Slicer

Prusa Slicer

### Host Software

OctoPrint

### Additional information & file uploads

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/7373999/Configuration.zip)

**Comment by thisiskeithb** [comment]
Whenever there are homing or leveling issues, we now ask everyone to follow a standard procedure to gather more information:

- Download Marlin [`bugfix-2.0.x`](https://github.com/MarlinFirmware/Marlin/archive/bugfix-2.0.x.zip) to test with the latest code.
- Enable `DEBUG_LEVELING_FEATURE` and `M114_DETAIL` and re-flash the firmware.
- Connect to your printer from host software such as Cura, Printrun or Repetier Host.
- Send `M502` and `M500` to ensure your Configurations are applied.
- Issue the command `M111 S247` to enable maximum logging.
- Perform a `G28` to do your standard homing procedure.
- Do a `G29` to probe the bed. This will also enable bed leveling.
- Do some of the moves that revealed problems before. Take notes.
- Copy the log output into a **.TXT file** and attach it to your next reply.

Repeat this procedure, if needed, to demonstrate inconsistencies. From these logs we should hopefully get a better idea of what's going on with your machine.

**Comment by bbrandst** [comment]
Attached is the log file containing the output of two complete sequences of the commands
M502, M500, M111 S247, G28, G29

[serial.txt](https://github.com/MarlinFirmware/Marlin/files/7375905/serial.txt)

I also made some recordings of the values delivered by the strain gauge and the ADC, respectively. It shows that the pressure on the nozzle (blue line) keeps increasing for 4-5 additional cycles after the probe has triggered (red line). The ADC delivers 10 readings per second and the probe triggers when the current ADC reading overshoots the average of the last 8 readings by 2000 or more.

![Measurement](https://user-images.githubusercontent.com/33779981/137973326-9a0d2bfa-72c3-467e-9069-90235844f2ec.png)

**Comment by thinkyhead** [comment]
> The nozzle continues moving down for an additional ~0.4 mm (or ~0.4 s) after the probe has triggered.

That's not quite possible, because Marlin aborts a homing move at the exact moment that it sees a probe trigger signal. There is no delay possible. So, if the axis is still moving, the signal has not yet arrived.

Your configuration looks okay, but I there is an extra setting `Z_PROBE_PAUSE_INTERVAL_MS`. What is that used for?

Try placing something under the bed to prevent the springs from springing (or use a temporary substitute bed made of a marble slab) and see if the strain gauge responds more quickly.

**Comment by bbrandst** [comment]
The extra setting `Z_PROBE_PAUSE_INTERVAL_MS` is there because I have inserted a piece of code as described in the RepRap forum thread [Precision Piezo Z-probe Now available](https://reprap.org/forum/read.php?1,767998,885356#msg-885356). It adds a pause of configurable length before probing, allowing the strain gauge to settle to avoid false triggers. This does not cause the problem, though, because the behavior was already the same before I added it.

Looking at the code I tend to agree with you that the movement should stop immediately. But I wonder what on earth could cause a delay of _several hundred_ milliseconds between the probe triggering and the signal reaching the handler.

The strain gauge _does_ respond quickly. This can be seen in the graph in my previous post and in addition there's also an LED connected to the trigger line, so I can see that it triggers as soon as the nozzle touches the bed.

Anyway, I'll try if making the bed more rigid changes anything.

**Comment by thinkyhead** [comment]
So something like this is added…
```cpp
#ifdef Z_PROBE_PAUSE_INTERVAL_MS
  if (Z_PROBE_PAUSE_INTERVAL_MS > 0) {
    millis_t time = millis() + Z_PROBE_PAUSE_INTERVAL_MS;
    while (millis() < time) idle();
  }
#endif
```
It seems like it should be fine, as described. But this is slightly neater…
```cpp
#if Z_PROBE_PAUSE_INTERVAL_MS > 0
  safe_delay(Z_PROBE_PAUSE_INTERVAL_MS);
#endif
```

**Comment by bbrandst** [comment]
Thanks for the tip with `safe_delay`, that's indeed quite a bit neater :-)

OK, I can confirm that the problem is not in Marlin. I used my oscilloscope to inspect the probe and stepper signals:
![DS1Z_QuickPrint1](https://user-images.githubusercontent.com/33779981/138557076-ec3ac41c-2074-479e-b67b-f4a07dbf819e.png)
The first two traces (yellow and cyan) are the DIR and STEP signals to one of the Z stepper drivers, and the third one (magenta) is the probe output. (It isn't really that noisy, I just used a rather lengthy unshielded wire to connect it to the oscilloscope probe.) Marlin does indeed immediately cease to step once the probe triggers.

I still don't understand how this correlates with my other measurements, but it seems the problem is in my probe.

Thus, I closed the issue, and sorry for the noise.
